### ISSUE TITLE
[PSyIR] Comparing datatypes

### ISSUE THREAD

**Issue by rupertford** [stfc/PSyclone#744]
The approach to comparing datatypes in #728 is not nice. In #728 there was a discussion about how to improve this. This issue is the result of that discussion.

**Comment by rupertford** [comment]
In #728 there were three options discussed

1) perform equality tests in the DataType classes
2) perform equality tests in a DataNode subclass of the relevant nodes (e.g. Literal)
3) use eq ne in the DataType classes to see if they are the same or not.

The problem with 1) is that (as far as I can tell) the base class would need to know about all of its subclasses to do the testing.

The problem with 2) is that a) it requires the datatype to already be attached to the node and b) it does not provide the methods to the DataSymbol class as it is not a Node. One solution to b) would be for DataNode and DataSymbol to use multiple inheritance.

The problem with 3) is that it only supports exact matching of datatypes and we might want to know if the datatype names (e.g. INTEGER) but not care about the precision. However this could be used in conjunction with the other options.

**Comment by rupertford** [comment]
If we go for option 2 then the code could look something like the following ...

```
    def is_scalar(self, name=None, precision=None):
        '''Return True if the node is scalar and False otherwise. The optional
        name argument additionally checks whether the type-name
        matches and the optional precision argument additionally
        checks whether the precision matches.

        :param name: optional argument used to check whether this \
            instance has the same type name as the argument.
        :type name: :py:class:`psyclone.psyir.symbols.ScalarType.Name`

        :param precision: optional argument used to check whether this \
            instance has the same precision as the argument.
        :type name: :py:class:`psyclone.psyir.symbols.ScalarType.Precision`, \
            int or :py:class:`psyclone.psyir.symbols.DataSymbol`

        '''
        name_match = True
        if name:
            name_match = self._datatype.name == name
        precision_match = True
        if precision:
            precision_match = self._datatype.precision == precision

        return (name_match and precision_match and
                isinstance(self._datatype, ScalarType))
```

**Comment by sergisiso** [comment]
> The problem with 1) is that (as far as I can tell) the base class would need to know about all of its subclasses to do the testing.

I don't see this. Consider the implementation below:
```
class DeferredType:
	def __eq__(self, other):
		return isinstance(other, DeferredType)

class ScalarType:
	def __eq__(self, other)
		return isinstance(other, ScalarType) and self.name == other.name \
                           and self.precision == other.precision

class ArrayType:
	def __eq__(self, other)
		return isinstance(other, ArrayType) and self.shape == other.shape \
                           and self.datatype == other.datatype
```
But I agree this just works for exact matching and we have lots of cases where we want a more flexible comparison mechanism.

**Comment by sergisiso** [comment]
Or maybe that is what you meant as option 3?

**Comment by rupertford** [comment]
Yes that was my option 3. I agree with adding the code as you suggest but as you say, we want a more flexible comparison scheme as well.

**Comment by hiker** [comment]
@sergisiso pointed out that I had a problem related to this, so I'll just add this here as an FYI:
```
print(create_arg_list.psyir_arglist[2].datatype.precision, psyir.R_DEF,
    create_arg_list.psyir_arglist[2].datatype.precision == psyir.R_DEF)
    r_def: DataSymbol<Scalar<INTEGER, UNDEFINED>, Import(container='constants_mod')>
    r_def: DataSymbol<Scalar<INTEGER, UNDEFINED>, Import(container='constants_mod')> False
```
So basically, the two data symbols (`create_arglist...precision` and `psyir.R_DEF`) are identical (both integer, no precision, imported from constants_mod), but the comparison still gives ``False``. For now I just compare the name of `...datatype.precision` with `r_def`. I've added a #TODO in for `kern_call_arg_list_test.py` in ffed25c98eb02f782a31c295e76afffac0577590

**Comment by arporter** [comment]
Googling, I found:
![image](https://github.com/stfc/PSyclone/assets/17409212/a31a4ba9-ad25-4306-b4cd-2d6e0790a84f)
(from https://eng.lyft.com/hashing-and-equality-in-python-2ea8c738fb9d). Since our Symbol class *is* mutable and (in a branch) implements `__eq__`, it cannot be hashable.

Python's `obj is in obj_list` uses `__eq__`.

**Comment by arporter** [comment]
I've pushed what I have so far to the `2223_datasym_equality` branch but I'm thinking that this might not be the best way to go. I'll return to the work that triggered this (#2197) and see if I can re-generate the use-case for this functionality.
